src/interfaces/plugins.ts

```diff
@@ -13,45 +13,114 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
 */
-
-const debug = require('debug')('signalk:interfaces:plugins')
-const fs = require('fs')
-const path = require('path')
-const express = require('express')
-const _ = require('lodash')
+import Debug from 'debug'
+import { Request, Response } from 'express'
+const debug = Debug('signalk:interfaces:plugins')
+// @ts-ignore
+import { getLogger } from '@signalk/streams/logging'
+import express from 'express'
+import fs from 'fs'
+import _ from 'lodash'
+import path from 'path'
+
+// tslint:disable-next-line:no-var-requires
 const modulesWithKeyword = require('../modules').modulesWithKeyword
-const getLogger = require('@signalk/streams/logging').getLogger
-const _putPath = require('../put').putPath
-const { getModulePublic } = require('../config/get')
-const { queryRequest } = require('../requestResponse')
+// tslint:disable-next-line:no-var-requires
+const put = require('../put')
+// tslint:disable-next-line
+const _putPath = put.putPath
+// tslint:disable-next-line:no-var-requires
+const getModulePublic = require('../config/get').getModulePublic
+// tslint:disable-next-line:no-var-requires
+const queryRequest = require('../requestResponse').queryRequest
 
 // #521 Returns path to load plugin-config assets.
 const getPluginConfigPublic = getModulePublic('@signalk/plugin-config')
 
-const DEFAULT_ENABLED_PLUGINS = process.env['DEFAULTENABLEDPLUGINS']
-  ? process.env['DEFAULTENABLEDPLUGINS'].split(',')
+const DEFAULT_ENABLED_PLUGINS = process.env.DEFAULTENABLEDPLUGINS
+  ? process.env.DEFAULTENABLEDPLUGINS.split(',')
   : []
 
-module.exports = function (app) {
-  const onStopHandlers = {}
+export type PluginFactory = (serverApi: ServerAPI) => Plugin
+
+export interface Plugin {
+  start: (config: any, restart: (newConfiguration: any) => void) => any
+  stop: () => any
+}
+
+interface PluginInfo extends Plugin {
+  enableLogging: any
+  packageName: any
+  packageLocation: string
+  registerWithRouter: any
+  signalKApiRoutes: any
+  name: string
+  id: string
+  schema: () => void | object
+  uiSchema: () => void | object
+  version: string
+  description: string
+  state: string
+  enabledByDefault: boolean
+  statusMessage: () => string | void
+}
+
+export interface ServerAPI {
```

Comment:
Yep, we never should have published getStream/Bus via streambundle. Need to find a way to keep backwards compatibility (via `.streambundle`)  and probably add new methods for them in the server api. Same thing for `.subscriptionmanager`.

`registerDeltaInputHandler` is really meant for intercepting and filtering/altering incoming data, not for getting input.

I need to look also into integrating the docs with the ts code now that we will have a clear API definition.

Reply:
As with my previous comment, re: `doRegisterPlugin()` line:
`const appCopy: ServerAPI = _.assign({}, app, { getSelfPath, getPath, ...`

Isn't `.streambundle` and `.subscriptionmanager` also copied to `appCopy` and thus already available to legacy plugins?

Re: docs - I suggest having the interface be decorated with JSDoc comments, and adding a JSDoc run be part of the server build with the resulting HTML files being checked in to the project along with source code.  That way, API docs are always updated automatically.  That is what we do in the [SensESP project](https://github.com/SignalK/SensESP/blob/master/README.md).  The "README.md" just has a link to our API page, and the API page is a set of HTML pages generated from a Doxygen run (Doxygen being the C++ equiv of JSDoc).

Reply:
> `registerDeltaInputHandler` is really meant for intercepting and filtering/altering incoming data, not for getting input.

So, which method is used to get the delta streams for processing?  Is it `registerActionHandler()`?

Reply:
No, BaconJS streams via `streambundle`.

Reply:
> No, BaconJS streams via `streambundle`.

Ok. If you for sure want to keep the BaconJS streams, I'd vote for simple `getBus()/Stream` methods directly in the interface, and define it in the same way you defined `registerDeltaInputHandler()` as a "self unsubscribing" call. Anything you can do to make things simpler for a plugin developer to come up to speed, as well as prevent them from making a mistake like introducing a resource leak.

That being said - do you want to endorse BaconJS, which is a very Javascript specific technique?  I saw some discussions on the SignalK Specifications page where there was some debate with the developer of the Artemis Server about maybe making a plugin API that is shareable between the two servers, perhaps even supporting different scripting languages.  I don't know how important something like that is to the overall vision. (For what its worth, I really like working with the BaconJS API, so if it is in fact endorsed as a preferred method, I'm all for it!)

Reply:
> I suggest having the interface be decorated with JSDoc comments

BTW - I will volunteer to do some of this grunt work if you would like (adding JSDoc comments, setting up a JSDoc run as part of the build, etc.).  Would it be beneficial to start a GitHub branch and apply this PR to that branch?  I could then fork the project and work on the branch, sending you PRs to the branch.

Reply:
For the time being I value js friendliness over portability for plugins. Plugins are more than an API: the app store/registry and the UI.

I just pushed 0eb123ece810ae69c89478350aeb92e38cbb23a8 with typedoc support - jsDoc equivalent for TypeScript. Give it a spin, you can fetch from upstream and checkout this branch.

I look forward to your help with all of this. I would rather merge the ts support quickly and continue working on the docs and details then - infrastructure I would like to have in place sooner rather than later. I probably won't have any meaningful capacity for handling PRs for next 2 weeks.